pkg/capabilities/consensus/ocr3/datafeeds/feeds_aggregator.go

```diff
@@ -37,23 +41,24 @@ type dataFeedsAggregator struct {
 
 var _ types.Aggregator = (*dataFeedsAggregator)(nil)
 
-// EncodableOutcome is a list of AggregatedPricePoints
+// This Aggregator has two phases:
+//  1. Agree on valid trigger signers by extracting them from event metadata and aggregating using MODE (i.e. F+1 copies needed).
+//  2. For each FeedID, select latest valid report, using signers list obtained in phase 1.
+//
+// EncodableOutcome is a list of aggregated price points.
 // Metadata is a map of feedID -> (timestamp, price) representing onchain state (see DataFeedsOutcomeMetadata proto)
-func (a *dataFeedsAggregator) Aggregate(previousOutcome *types.AggregationOutcome, observations map[ocrcommon.OracleID][]values.Value) (*types.AggregationOutcome, error) {
-	// find latest valid Mercury report for each feed ID
+func (a *dataFeedsAggregator) Aggregate(previousOutcome *types.AggregationOutcome, observations map[ocrcommon.OracleID][]values.Value, f int) (*types.AggregationOutcome, error) {
+	allowedSigners, minRequired, payloads := a.extractSignersAndPayloads(observations, f)
+	if len(payloads) > 0 && minRequired == 0 {
```

Comment:
this is an odd check. why not return an error from the extract? or maybe rename minRequired?

Reply:
It's a little tricky because for empty observations (from valid nodes!) we will extract zero minRequired but we still want to complete the rest of Aggregate to return meaningful outcome. I re-worded the error message to be more descriptive.